### PR TITLE
[ADDED] Tiered limits and fix limit issues on updates

### PR DESCRIPTION
**Pull request by matthiashanel**
Signed-off-by: Matthias Hanel <mh@synadia.com>

This keeps proper track of reservations (computes them on the fly)
stream updates now pass when at the stream limit. (Similar for changes in replication factor)
Adding tiers also influenced the layout of the info message.

### CODE REVIEW

File: server/jetstream.go [hunk 1]
```diff
@@ -1567,13 +1686,36 @@ func (jsa *jsAccount) sendClusterUsageUpdate() {
 	}
 	jsa.lupdate = now
 
-	b := make([]byte, 32)
+	lenUsage := len(jsa.usage)
```
Comment (kozlovic) [inline server/jetstream.go:1689]: Have you verified that it will work well with older versions of the server?

**Review by kozlovic — CHANGES_REQUESTED** [review]
First pass.. would need to review a bit more carefully jetstream.go (did that one last and ran out of steam ;-) )

File: server/stream.go [hunk 2]
```diff
@@ -313,11 +314,25 @@ func (a *Account) addStreamWithAssignment(config *StreamConfig, fsConfig *FileSt
 			return nil, ApiErrors[JSStreamNameExistErr]
 		}
 	}
-	// Check for account and server limits.
-	if err := jsa.checkAllLimits(&cfg); err != nil {
-		jsa.mu.Unlock()
+	selected, tier, hasTier := jsa.selectLimits(&cfg)
+	reserved := int64(0)
+	if js.cluster == nil {
```
[the remark this answers is not in the document]
Comment (kozlovic) [reply server/stream.go:319]: After further investigation, js.cluster is immutable so it should be safe to keep your code like that.

File: server/stream.go [hunk 3]
```diff
@@ -984,21 +1006,47 @@ func (jsa *jsAccount) configUpdateCheck(old, new *StreamConfig) (*StreamConfig,
...
+	// Check limits.
+	jsa.mu.RLock()
+	js := jsa.js
+	acc := jsa.account
+	selected, tier, hasTier := jsa.selectLimits(&cfg)
+	if !hasTier && old.Replicas != cfg.Replicas {
+		selected, tier, hasTier = jsa.selectLimits(old)
+	}
+	reserved := int64(0)
+	if js.cluster == nil {
```
[the remark this answers is not in the document]
Comment (kozlovic) [reply server/stream.go:1031]: After further investigation, js.cluster is immutable so it should be safe to keep your code like that.

File: server/stream.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthiashanel) [reply server/stream.go:319]: it is immutable and should not matter. however, it does not seem to be used that way anywhere else...
Also I can switch to a readlock

**Review by kozlovic — APPROVED** [review]
LGTM

File: server/jetstream.go [hunk 4]
```diff
@@ -54,6 +54,7 @@ type JetStreamStats struct {
 	ReservedMemory uint64            `json:"reserved_memory"`
 	ReservedStore  uint64            `json:"reserved_storage"`
 	Accounts       int               `json:"accounts"`
+	RaftAssets     int               `json:"raft_assets"`
```
Comment (derekcollison) [inline server/jetstream.go:57]: I think Ben was already tracking this so need to alert him. Also, might move away from raft hence why I called them HA vs Raft, so let's put that back.

**Review by derekcollison — APPROVED** [review]
LGTM

File: server/jetstream.go (hunk 4 above)
Comment (matthiashanel) [reply server/jetstream.go]: reverted

**Pull request merged**